# Demand and Supply factors impacted home prices over the last 20 years
Submitted by,
Avinash A Godi

# Used Economic Factors
## Supply Side Factors That Affect Home Prices
Various factors that affect the supply of homes available for sale are discussed below:

## Months of supply is the basic measure of supply itself in the real estate market 

Data Source:
Months of Supply: https://fred.stlouisfed.org/graph/?g=zneA

## Unemployment
Unemployment can also affect both demand and supply in the real estate industry.
A high unemployment rate can mean that people simply do not have the money to spend on houses.
It can also mean that there is lower investment in the industry and hence lower supply.

Data Source: https://fred.stlouisfed.org/series/UNRATE

## Mortgage Rate
Mortgage rates are a huge factor that decide how well the real estate market will perform.
It plugs into both supply and demand side of the equation. It affects the availability of financing options to buyers,
as well as the ease of financing new constructions.
It also affects the delinquency rate and the number of refinances for mortgages.
People are more likely to default on a higher mortgage rate!

Data Source: https://fred.stlouisfed.org/graph/?g=zneW

## USA GDP
The GDP is a measure of output of the economy overall, and the health of the economy.
An economy that is doing well usually implies more investment and economic activity, and more buying.

Data Sources:
Monthly Index: https://fred.stlouisfed.org/graph/?g=znfe

## Demand Side Factors That Affect Home Prices
Demand for housing, and specifically, home ownership, is affected by many factors, some of which are closely inter-related.
Many of these factors also affect the supply in housing market.
Below are a few factors that are prominent in influencing the demand for home buying:

## Affordability: Wages & Disposable Personal Income
The measure is disposable personal income: how much of the earning is actually available to an individual for expenditure.
This is an important measure as well, as it takes into account other factors like taxes etc.

Data Sources:
Real Disposable Personal Income: https://fred.stlouisfed.org/series/DSPIC96#0

## Personal Savings
The extent to which people are utilizing their personal income for savings matters in overall investments and capital
availability, and the interest rate for loans (and not just the mortgage rate).
It is also an indicator of how much the current population is inclined to spend their money, vs save it for future use.
This is an indicator of the demand for home ownership as well.

Data Source:
Personal Saving: https://fred.stlouisfed.org/series/PMSAVE

## Behavioural Changes & Changes in Preferences
Changes in home ownership indicate a combination of factors including change in preferences and attitudes of people towards
home buying. Change in cultural trends can only be captured by revealed preferences, and this metric can be taken as a
revealed metric for propensity for home buying.

The other metric to track changes in preferences is personal consumption expenditure. For eg, if expenditure is increasing,
but there is no such increase in homeownership, it would indicate a change in preferences towards home buying and ownership.
Maybe people prefer to rent a home than buying one. Hence, both of these parameters are used.

Date Sources:
Personal Consumption Expenditures: https://fred.stlouisfed.org/series/PCE

In [151]:
# Reading the csv and excel file from the above links
import pandas as pd
import numpy as np

In [152]:
# Months of supply
MOS=pd.read_csv("Months-of-supply-MSACSR.csv")
MOS

DATE  MSACSR
0    2002-07-01     4.2
1    2002-08-01     4.0
2    2002-09-01     3.9
3    2002-10-01     4.0
4    2002-11-01     4.0
..          ...     ...
236  2022-03-01     7.0
237  2022-04-01     8.4
238  2022-05-01     8.4
239  2022-06-01     9.2
240  2022-07-01    10.9

[241 rows x 2 columns]

In [153]:
#unemployement
Unem=pd.read_csv("Unemployement-Rate-UNRATE.csv")
Unem

DATE  UNRATE
0    2002-07-01     5.8
1    2002-08-01     5.7
2    2002-09-01     5.7
3    2002-10-01     5.7
4    2002-11-01     5.9
..          ...     ...
236  2022-03-01     3.6
237  2022-04-01     3.6
238  2022-05-01     3.6
239  2022-06-01     3.6
240  2022-07-01     3.5

[241 rows x 2 columns]

In [154]:
#Mortgage Rate
MR=pd.read_csv("Mortgage-Rate-MORTGAGE30US.csv")
MR

DATE  MORTGAGE30US
0    2002-07-01        6.4850
1    2002-08-01        6.2900
2    2002-09-01        6.0925
3    2002-10-01        6.1125
4    2002-11-01        6.0680
..          ...           ...
236  2022-03-01        4.1720
237  2022-04-01        4.9825
238  2022-05-01        5.2300
239  2022-06-01        5.5220
240  2022-07-01        5.4125

[241 rows x 2 columns]

In [155]:
#USA GDP
GDP=pd.read_csv("USA-GDP-USALORSGPNOSTSAM.csv")
GDP

DATE  USALORSGPNOSTSAM
0    01-07-2002         98.937697
1    01-08-2002         98.836909
2    01-09-2002         98.723374
3    01-10-2002         98.608223
4    01-11-2002         98.504619
..          ...               ...
236  01-03-2022         99.916488
237  01-04-2022         99.754563
238  01-05-2022         99.596427
239  01-06-2022         99.827605
240  01-07-2022         99.827605

[241 rows x 2 columns]

In [156]:
GDP['DATE'] = pd.to_datetime(GDP['DATE'], errors='coerce')

In [157]:
GDP['DATE']=GDP['DATE'].dt.strftime('%Y-%d-%m')

In [158]:
GDP

DATE  USALORSGPNOSTSAM
0    2002-07-01         98.937697
1    2002-08-01         98.836909
2    2002-09-01         98.723374
3    2002-10-01         98.608223
4    2002-11-01         98.504619
..          ...               ...
236  2022-03-01         99.916488
237  2022-04-01         99.754563
238  2022-05-01         99.596427
239  2022-06-01         99.827605
240  2022-07-01         99.827605

[241 rows x 2 columns]

In [159]:
#Real Disposal Income
RDI=pd.read_csv("Real-Disposal-Income-DSPIC96.csv")
RDI

DATE  DSPIC96
0    2002-07-01  10053.5
1    2002-08-01  10056.6
2    2002-09-01  10068.3
3    2002-10-01  10097.3
4    2002-11-01  10124.3
..          ...      ...
236  2022-03-01  15119.6
237  2022-04-01  15143.3
238  2022-05-01  15142.2
239  2022-06-01  15105.1
240  2022-07-01  15145.6

[241 rows x 2 columns]

In [160]:
#Personal Savings
PS=pd.read_csv("Personal-Savings-PMSAVE.csv")
PS

DATE  PMSAVE
0    2002-07-01   441.7
1    2002-08-01   435.9
2    2002-09-01   482.0
3    2002-10-01   474.3
4    2002-11-01   476.6
..          ...     ...
236  2022-03-01   961.2
237  2022-04-01   949.1
238  2022-05-01   964.2
239  2022-06-01   921.7
240  2022-07-01   932.3

[241 rows x 2 columns]

In [161]:
#Personal Consumption Expenditure
PCE=pd.read_csv("Personal-Consumption-Expenditure-PCE.csv")
PCE

DATE      PCE
0    2002-07-01   7387.3
1    2002-08-01   7412.6
2    2002-09-01   7391.5
3    2002-10-01   7435.2
4    2002-11-01   7463.8
..          ...      ...
236  2022-03-01  16831.2
237  2022-04-01  16904.9
238  2022-05-01  16991.6
239  2022-06-01  17160.6
240  2022-07-01  17184.3

[241 rows x 2 columns]

In [162]:
#Housing Price Index y variable(S&P-Case-Shiller)
SP=pd.read_csv("S&P Case-shiller-CSUSHPISA.csv")
SP

DATE  CSUSHPISA
0    01-07-2002    122.888
1    01-08-2002    123.831
2    01-09-2002    124.780
3    01-10-2002    125.734
4    01-11-2002    126.669
..          ...        ...
236  01-03-2022    296.362
237  01-04-2022    301.409
238  01-05-2022    305.179
239  01-06-2022    306.181
240  01-07-2022    306.192

[241 rows x 2 columns]

In [163]:
#joining all Tables using joins and creating new x and y variable tables
x1=pd.merge(MOS,Unem,how='outer', on='DATE')

In [164]:
x2=pd.merge(x1,MR,how='outer',on='DATE')

In [165]:
x3=pd.merge(x2,GDP,how='outer',on='DATE')

In [166]:
x4=pd.merge(x3,RDI,how='outer',on='DATE')

In [167]:
x5=pd.merge(x4,PS,how='outer',on='DATE')

In [168]:
x_monthly=pd.merge(x5,PCE,how='outer',on='DATE')

In [169]:
x_monthly.head()

DATE  MSACSR  UNRATE  MORTGAGE30US  USALORSGPNOSTSAM  DSPIC96  \
0  2002-07-01     4.2     5.8        6.4850         98.937697  10053.5   
1  2002-08-01     4.0     5.7        6.2900         98.836909  10056.6   
2  2002-09-01     3.9     5.7        6.0925         98.723374  10068.3   
3  2002-10-01     4.0     5.7        6.1125         98.608223  10097.3   
4  2002-11-01     4.0     5.9        6.0680         98.504619  10124.3   

   PMSAVE     PCE  
0   441.7  7387.3  
1   435.9  7412.6  
2   482.0  7391.5  
3   474.3  7435.2  
4   476.6  7463.8

In [170]:
y_monthly=SP
y_monthly

DATE  CSUSHPISA
0    01-07-2002    122.888
1    01-08-2002    123.831
2    01-09-2002    124.780
3    01-10-2002    125.734
4    01-11-2002    126.669
..          ...        ...
236  01-03-2022    296.362
237  01-04-2022    301.409
238  01-05-2022    305.179
239  01-06-2022    306.181
240  01-07-2022    306.192

[241 rows x 2 columns]

# Building The Model
The S&P Case-Shiller Housing Price Index is taken as the y variable, or dependent variable, as an indicator of change in prices.
All the above factors make up independent variable or XN, the X vector as a transpose, in a multivariable regression model of order N, comprising X1, X2, X3…XN etc. In a typical prediction framework, how a variable affects the dependent variable is not relevant; only that it does or does not affect it is what  matters. As such, many of these factors are possibly endogenous and/or affect the prices in multiple ways: that mechanism is not the subject of study in this model. We are currently seeing the overall effect of the variables on price trends.

The regression itself does not run on time-series data, so the datetime columns are removed in the final data for the regression.

Multiple Linear Regression for Prediction
Exploratory Data Analysis is required to assess what type of regression is needed for building the model. Although using Linear or Multiple Linear Regression is traditional, it is still important to first look at the data and how it is spread out. It also helps eliminate variables which are closely correlated and redundant.

In [171]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn import metrics

x = x_monthly.drop(['DATE'],axis=1) #Removing date column

y = y_monthly.drop(['DATE'],axis=1) #Removing date column

x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.35, shuffle=False, stratify=None)

reg = LinearRegression()
reg.fit(x_train, y_train)

y_predict = reg.predict(x_test)

# Predicted Prices Vs Actual Prices
There are a few diagnostic measures to check if the model has worked well or not: checking the R2 coefficient and the root mean squared error are two such methods.

Checking key parameters for diagnostics:

In [172]:
print(reg.coef_)
print(reg.intercept_)
print(reg.score(x_train, y_train))
print(reg.score(x_test, y_test))

print('Mean Absolute Error:', metrics.mean_absolute_error(y_test, y_predict))
print('Mean Squared Error:', metrics.mean_squared_error(y_test, y_predict))
print('Root Mean Squared Error:', np.sqrt(metrics.mean_squared_error(y_test, y_predict)))

print(y_test.describe())

[[-1.297395   -2.58289052  6.11718308 -1.72249303  0.07616979 -0.09953122
  -0.02497553]]
[-254.55087112]
0.9189183869444802
-2.0495851734342296
Mean Absolute Error: 40.76279854143108
Mean Squared Error: 3952.801063394495
Root Mean Squared Error: 62.87130556457767
        CSUSHPISA
count   85.000000
mean   215.703529
std     36.216120
min    172.122000
25%    188.787000
50%    206.295000
75%    229.409000
max    306.192000


# Therefore from the above model I have achieved 
### Mean Absolute Error: 40.76279854143108
### Mean Squared Error: 3952.801063394495
### Root Mean Squared Error: 62.87130556457767